### ISSUE TITLE
Device.getShaderPrecisionFormat(shaderType, precisionType)

### ISSUE THREAD

**Issue by cerisano** [gpuweb/gpuweb#1048]
Since GLES 2.0, WebGL has been able to [determine apriori the exact hardware supported precision](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGetShaderPrecisionFormat.xml) of numeric types in shaders.
This an essential requirement for doing any rendering or computation that has precision dependencies.

The Mozilla WebGL documentation for this is given [here](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/getShaderPrecisionFormat).

WebGPU of course has the same requirement, especially given that it exposes compute pipelines where precision dependencies will very much determine the quality of the results. 

Without such a method to ensure precision aprori, WebGPU (and WebGL) will return non-deterministic results on various hardware (eg. mobile vs internal vs discrete GPU make/models).

Would like to suggest and additional method to the Device class with a pseudo signature similar to WebGL as the following:

`ShaderPrecisionFormat Device.getShaderPrecisionFormat(shaderType, precisionType)`

Where paramaters are as:

- `shaderType` is vertex, fragment or compute.
- `precisionType` is any of the numeric types (most specifically floats).

And where return value is as:

- ShaderPrecisionFormat =  {max_value, min_value, precision}
- precision is the exact number of bits the hardware supports for a given numeric type.

 I had considered raising this requirement for a precision check in WGSL, but there is already precedent set in WebGL.

**Comment by kainino0x** [comment]
WGSL decided not to have relaxed precision, so I'm pretty sure all floats are 32-bit. However, I believe the precision of every single operation (arithmetic, trigonometric, etc.) could vary to some degree, and I'm not sure how to deal with that.

**Comment by cerisano** [comment]
It is a hardware query, to determine what precision is supported on the metal.

**Comment by kainino0x** [comment]
IIUC: Since there are no relaxed precisions, there are no lowp, mediump, and highp to query. There's only a single float precision of 32 bits, with precision loss only allowed inside math operations.

**Comment by kainino0x** [comment]
Decision on that was https://github.com/gpuweb/gpuweb/issues/658#issuecomment-613612144

**Comment by cerisano** [comment]
I am only suggesting querying the exact number bits supported for floats in the hardware (low, med, hi are just ranges),

That is what will cause precision loss in math operations (eg. iPhones are only fp16), and there needs to be a way to at least check for that apriori.

**Comment by kainino0x** [comment]
I'm saying that I'm pretty sure that the query would always return "32". Unless I misunderstand non-relaxed precision or its hardware support. @dneto0 maybe can help?

**Comment by cerisano** [comment]
It is querying the hardware, not the API.

**Comment by kainino0x** [comment]
in my understanding, we require the hardware to support 32-bit, and we don't even allow access to lower precisions (unless a 16-bit float extension is enabled).

Hardware-wise, I believe this is required starting with OpenGL ES 3.0 (WebGL 2.0) and is what you get if you say `precision highp float;` at the top.

**Comment by cerisano** [comment]
So WebGPU will not run on iPhones, for example?

**Comment by kainino0x** [comment]
Since highp, mediump, lowp don't exist in WebGPU, there would be no `precisionType` in your suggested query signature.

**Comment by kainino0x** [comment]
iPhones support OpenGL ES 3.0. `float` in Metal Shading Language is exactly 32 bits.

**Comment by cerisano** [comment]
It ws only a suggested signature - call it getFloatingPointPrecision then (with no parameters).

**Comment by kainino0x** [comment]
It would always return 32.

**Comment by cerisano** [comment]
Not on an iPhone. It would return 16.

**Comment by kainino0x** [comment]
Page 208+ of the MSL spec specifies the required precision of all of the math ops/functions.

**Comment by kainino0x** [comment]
The MSL spec says `float` is 32 bits though. Is that not the precision you're trying to query?

**Comment by cerisano** [comment]
I want to know what the hardware supports. Having the API return 32 when the hardware only supports 16 is an error.

**Comment by kainino0x** [comment]
Can you provide a reference showing that the real precision is not 32 bits? The MSL spec seems pretty clear. Is there a way to query this in Metal?

**Comment by cerisano** [comment]
Some manufacturers optimize for fp16 for performance and power consumption reasons. It don't see how WebGPU can guarantee 32 bit precision on all devices.

https://www.realworldtech.com/apple-custom-gpu/

**Comment by cerisano** [comment]
It seems that WebGPU can't rely on implementions to actually perform fp32 operations in a non-optimized manner (eg. the device converting back and forth from fp16 to fp32 for performance reasons).

**Comment by kainino0x** [comment]
The table of math op precisions I mentioned, on page 208, shows that the results of math ops are required to be within a few 32-bit-ULP of the exactly-correct result.

**Comment by cerisano** [comment]
That presupposes the device is reporting 32 bit precision, when in fact it might be optimized/augmented by fp16 ops under the hood.

**Comment by kainino0x** [comment]
I didn't see anything in the MSL spec that says the device is allowed to report precisions at all. It says `float` is 32 bits and `half` is 16 bits. AFAIK you can always use those, with those precisions, on any hardware that supports Metal.

Additionally, I really don't think that precisions higher than 32 bits are allowed. For example, the minimum accuracy of `x * y` is "Correctly rounded". That's taking RealNumberValue(f32 x) times RealNumberValue(f32 y) and then rounding it to f32 with exact precision. I think you could get a different result if x and y were secretly more precise.

**Comment by kainino0x** [comment]
> Implementations might pass the MSL spec on some but not others, and not have reliable 32 bit precision.

This shouldn't be possible. If Apple's own implementations of MSL can't pass the MSL spec, then the MSL spec is wrong.

**Comment by cerisano** [comment]
I am refernig to the WebGPU spec that seems to guarantee 32 bits of precision wihout taking into account hardware that optimizes 32 bit floats under certain conditions (for performance, power consumption, etc).

**Comment by kainino0x** [comment]
If that's the case then WebGPU will have to spec and expose that precisions can be lower. Not allow implementations to be out of spec. Of course WebGPU will spec precisions for all the math operations, but that's different from saying float can be less than 32 bits. Again, if that's the case, we need some reference that says so.

The article you posted only says that 32-bit floats and 16-bit floats are both supported, 16-bit floats are more efficient in memory and in ops, and that conversions between the two are cheap to allow mixing of the two for manual optimization. Not that there are any automatic optimizations that reduce precision.

In response to the second thing: half floats are exactly 16-bit, not relaxed precision. Please take a look at the Metal Shading Language spec.

**Comment by cerisano** [comment]
Any "optimizing" of fp32 precision operations with fp16 precision operations would seem to reduce precision by half on the face of it. Is a reference really required?

Also I am not referring to MSL, Metal or any other spec. I am referring to the WebGPU API which needs be able to guarantee that floats and half-float operations will always be least 32 or 16 bits in precision (resp.) under all hardware conditions.

**Comment by Kangz** [comment]
All APIs we target have a way to express 32bit floats in their shading language, and AFAIK all of the APIs guarantee they conform to IEEE 754 with no precision loss. There isn't a single bit of precision loss for regular arithmetic float operations. There isn't conversion to FP 16 and back for arithmetic precisions. When you ask 32 bit floats, you get exactly that, 32bit IEEE 754 floats with correctly implemented arithmetic operations.

Where you can see variance in precision are in trigonometric and transcendental function (`sin`, `exp` etc) because they are implemented with iterative algorithms, lookup tables and other mechanisms in hardware. This is most likely the issue you were seeing in your ray-tracing demo. All APIs specify minimum precisions to reach with this operations in number of ULP so there isn't a query that can say "this is 16bit or 32bit". The computations are all 32bit but how precise the result is varies a bit depending for example on how far the hardware computes the Taylor serie.

**Comment by kainino0x** [comment]
We discussed this on the WebGPU-NAPI call, and I think we came to an agreement on getShaderPrecisionFormat not being useful here. Instead, what applications might want to know is the hardware precisions of individual operations. However, I don't think the native APIs give us this information, so it doesn't make sense to expose queries in WebGPU - implementations would have to run a bunch of precision tests on startup.

Instead, I think what is needed is a community library that can execute these tests (run ops in various stages with various input values) and approximate the actual precision of individual operations. This would give the app the precision info it may need, as well as allow manually exploring the results on various devices.

Another thing to note is that we'll eventually have conformance tests for math op precision, which run each op with many different input values and ensures that all of the results are within the precision we guarantee in the WGSL spec. While of course we can't change the hardware precision, this will serve two purposes:
- Ensure that we don't accidentally have fast-math or relaxed precision enabled in implementations
- Ensure that the maximum-error values in the spec actually reflect real hardware
